### PR TITLE
feat: add JARPATTERN env var to java runtimes to identify binary

### LINKED ISSUE #169: Fix runtime support
Currently, only Spring Boot runtime works correctly. This is due to the fact that we've been experimenting with maven build from source on cluster. This requires a specific image for several reasons:
1. we wanted to have a first build that would be as fast as possible so we experimented with bundling an initial maven repository with the image. Of course, this only works if you know what theses dependencies should be. In the case of Spring Boot, we could pre-load all the dependencies defined in the Snowdrop BOM. We could certainly adopt a similar strategy for the other java runtimes but we just haven't done so yet.
2. we need to be able to perform several maven builds without having to re-download the dependencies all the time. There are several ways to accomplish this: either use docker layering to bundle the mostly unchanging dependencies or try to somehow persist the maven repository across pod restart / re-creation. We have experimented with the second option and as a consequence, this requires a specific image layout and persistent volume setup.
3. as the linking process results in the component pod being dropped and re-created, we also needed to be able to persist any binary that might have already been built so that the user wouldn't lose their app and needed to do another build after creating a link. Similarly, we also needed to be able to automatically restart any existing binary for the same reason. This also requires a specific run command and set up.

Theses reasons led to the creation of a [`hal-mvn-jdk`](https://github.com/halkyonio/container-images/tree/master/hal-mvn-jdk) image that's currently only being used by the Spring Boot runtime (since it bundles Spring Boot dependencies) and associated supervisor commands. As a result, the default s2i java image is not compatible anymore. 

We have 2 options to solve this issue: either create runtime-specific images with appropriate dependencies being bundled or make the `hal-mvn-jdk` image more generic by removing all the Spring Boot dependencies so that it's lighter (for a faster initial image pull) but with a slightly slower initial build.

Generifying the `hal-mvn-jdk` image is probably the best solution in the short term (and maybe longer term as well, since there are differing opinions on the value of bundling initial dependencies in the runtime image).

### PR DESCRIPTION
**Pull request by metacosm**
Fixes #169

### CODE REVIEW

**Comment by cmoulliard** [comment]
To be documented too under the README. @metacosm

**Comment by jponge** [comment]
Can you tweak the pattern from the resource manifest?

**Comment by jponge** [comment]
Not everybody prepares a shaded jar name `*-all.jar`, some people use `*-fat.jar`, etc.

**Comment by cmoulliard** [comment]
> I don't think that it should be documented here actually. This is documented at the image level which is where it needs to be, in my opinion.

As the operator owns the process to create the pods from the list of the images mapped with the runtime property and that could also certainly overrided this property using an ENV VAR, then that should be documented here.

**Comment by cmoulliard** [comment]
LGTM. Could be merged

File: pkg/controller/component/runtimes.go [hunk 1]
```diff
@@ -16,13 +16,13 @@ const (
...
-		"vert.x":          newRuntime("quay.io/halkyonio/hal-maven-jdk"),
-		"quarkus":         newRuntime("quay.io/halkyonio/hal-maven-jdk"),
-		"thorntail":       newRuntime("quay.io/halkyonio/hal-maven-jdk"),
-		"openjdk8":        newRuntime("registry.access.redhat.com/redhat-openjdk-18/openjdk18-openshift"),
-		"node.js":         newRuntime("nodeshift/centos7-s2i-nodejs"),
-		supervisorImageId: newRuntime("quay.io/halkyonio/supervisord"),
+		"spring-boot":     newRuntime("quay.io/halkyonio/hal-maven-jdk", "*"),
+		"vert.x":          newRuntime("quay.io/halkyonio/hal-maven-jdk", "*-all"),
+		"quarkus":         newRuntime("quay.io/halkyonio/hal-maven-jdk", "*-runner"),
+		"thorntail":       newRuntime("quay.io/halkyonio/hal-maven-jdk", "*"),
```
Comment (Ladicek) [inline pkg/controller/component/runtimes.go]: I think this could/should be `*-thorntail`.

Comment (metacosm) [reply pkg/controller/component/runtimes.go]: Indeed, though it doesn't really matter since the build only generates one jar… :)

Comment (Ladicek) [reply pkg/controller/component/runtimes.go]: It only generates one JAR if `<packaging>war</packaging>`. That is 99% of cases, probably, but people can have `<packaging>jar</packaging>`.

Comment (metacosm) [reply pkg/controller/component/runtimes.go]: Noted and fixed! Thank you.

**Pull request merged**